internal/pkg/mining/scheduler.go

```diff
@@ -117,39 +136,61 @@ func (s *timingScheduler) IsStarted() bool {
 	return s.isStarted
 }
 
-func nextNullBlkCount(prevNullBlkCount uint64, prevBase, currBase block.TipSet) uint64 {
-	if !prevBase.Defined() || !prevBase.Key().Equals(currBase.Key()) {
-		return 0
+// NewScheduler returns a new timingScheduler to schedule mining work on the
+// input worker.
+func NewScheduler(w Worker, f func() (block.TipSet, error), c clock.ChainEpochClock) Scheduler {
+	return &timingScheduler{
+		worker:       w,
+		pollHeadFunc: f,
+		chainClock:   c,
 	}
+}
 
-	return prevNullBlkCount + 1
+// MineOnce mines on a given base until it finds a winner.
+func MineOnce(ctx context.Context, w DefaultWorker, ts block.TipSet, c clock.ChainEpochClock) (Output, error) {
+	var winner *block.Block
+	var err error
+	var nullCount uint64
+	for winner == nil {
+		winner, err = MineOneEpoch(ctx, w, ts, nullCount, c)
+		if err != nil {
+			return Output{}, err
+		}
+		nullCount++
+	}
 
+	return Output{NewBlock: winner}, nil
 }
 
-// NewScheduler returns a new timingScheduler to schedule mining work on the
-// input worker.
-func NewScheduler(w Worker, md time.Duration, f func() (block.TipSet, error)) Scheduler {
-	return &timingScheduler{worker: w, mineDelay: md, pollHeadFunc: f}
-}
+// MineOneEpoch attempts to mine a block in an epoch and returns the mined block
+// or nil if no block could be mined
+func MineOneEpoch(ctx context.Context, w DefaultWorker, ts block.TipSet, nullCount uint64, chainClock clock.ChainEpochClock) (*block.Block, error) {
+	workCtx, workCancel := context.WithCancel(ctx)
+	defer workCancel()
+	outCh := make(chan Output, 1)
 
-// MineOnce is a convenience function that presents a synchronous blocking
-// interface to the mining scheduler.  The worker will mine as many null blocks
-// on top of the input tipset as necessary and output the winning block.
-// It makes a polling function that simply returns the provided tipset.
-// Then the scheduler takes this polling function, and the worker and the
-// mining duration
-func MineOnce(ctx context.Context, w Worker, md time.Duration, ts block.TipSet) (Output, error) {
-	pollHeadFunc := func() (block.TipSet, error) {
-		return ts, nil
+	// Control the time so that this block is always mined at a time that matches the epoch
+	h, err := ts.Height()
+	if err != nil {
+		return nil, err
 	}
-	s := NewScheduler(w, md, pollHeadFunc)
-	subCtx, subCtxCancel := context.WithCancel(ctx)
-	defer subCtxCancel()
+	epochStartTime := chainClock.StartTimeOfEpoch(nullCount + h + 1)
+	nextEpochStartTime := chainClock.StartTimeOfEpoch(nullCount + h + 2)
+	epochTime := nextEpochStartTime.Sub(epochStartTime)
+	halfEpochTime := epochTime / time.Duration(2) // mine blocks midway through the epoch
+
+	w.clock = th.NewFakeClock(epochStartTime.Add(halfEpochTime))
```

Comment:
This clock futzing is pretty nasty, but I think also avoidable. Now that you've removed the sleep from the worker, it only uses the clock to produce a timestamp. So just pass the timestamp to w.Mine().

Worker should probably be renamed to BlockProducer or similar, and just merge the Generate function into the same file (follow-up).

Reply:
We could do that but I don't think we should.  Right now the timestamp indicates when the block production finished by default.  We would lose that potentially useful signal by default if we timestamped all blocks in the beginning of the epoch.

The clock futzying is mildly ugly but really not so bad and not hit on the default production code path anyway.

I'm fine with moving generate into the same file as a low priority thing.